Fix reusable block cancelling

noisysocks opened the pull request:
![dog](https://user-images.githubusercontent.com/612155/37696086-100b0fdc-2d28-11e8-9617-0fd6c45418af.gif)

Makes it so that pressing Cancel while editing a reusable block discards any unsaved changes that have been made to that block.

This fixes #5719, which is a regression introduced by https://github.com/WordPress/gutenberg/pull/5228.

### To test:

1. Insert a reusable block
2. Select it and click Edit
3. Make some changes to the block
4. Click Cancel
5. The block should appear how it did before you edited it

noisysocks committed:
Fix reusable block cancelling

Pressing Cancel while editing a reusable block should discard any
unsaved changes that have been made to that block.

aduth commented on blocks/library/block/index.js:
```diff
@@ -30,6 +30,7 @@ class ReusableBlockEdit extends Component {
 		this.state = {
 			isEditing: !! ( reusableBlock && reusableBlock.isTemporary ),
 			title: null,
+			attributes: null,
```
I think the name of the state property could be improved to better indicate that it's the _transient_ updates which have yet to be saved. Otherwise it's a bit difficult to discern the difference between `this.state.attributes` and `this.props.block.attributes`.

noisysocks replied:
👌 fixed in 5a36e93ce.

aduth commented on blocks/library/block/index.js:
```diff
@@ -40,46 +41,53 @@ class ReusableBlockEdit extends Component {
+			if ( prevState.attributes !== null ) {
+				return { attributes: { ...prevState.attributes, ...attributes } };
+			}
+			return null;
```
Is an explicit return value necessary here? Or could we drop this line altogether?

noisysocks replied:
👌 fixed in 5a36e93ce.

aduth commented on blocks/library/block/index.js:
```diff
@@ -40,46 +41,53 @@ class ReusableBlockEdit extends Component {
 		this.setState( {
 			isEditing: true,
 			title: reusableBlock.title,
+			attributes: block.attributes,
```
Do we need to track all of the block's original attributes in state? Or could we simply track the state as the object which assigns into `block.attributes` as the base?

```js
this.setState( {
	attributes: {},
} );

// ... 

const attributes = { ...block.attributes, ...this.state.attributes };
```

Also has the added advantage that when we call `updateAttributes` in `save`, we're only applying the patch changes to state (likely to save a couple iterations when [aggregating into state value](https://github.com/WordPress/gutenberg/blob/030b562ed81b567806975a43900b18064ba12a8a/editor/store/reducer.js#L292) and maybe avoiding some false positives when [overriding state value for undo](https://github.com/WordPress/gutenberg/blob/030b562ed81b567806975a43900b18064ba12a8a/editor/store/reducer.js#L165)).

noisysocks replied:
👌 fixed in 5a36e93ce.

noisysocks committed:
Store changedAttributes instead of attributes

Avoid storing all of the referenced block's attributes in state by
storing only the changed attributes, and name this state property
changedAttributes to reflect this and to avoid ambiguity.
